pkg/api/persistentvolumeclaim/util_test.go

```diff
@@ -292,3 +293,164 @@ func TestAnyDataSourceFilter(t *testing.T) {
+			oldSpec:  getPVCSpec(),
+			expected: getPVCSpec(),
+		},
+		{
+			name:     "for:newPVC=hasfield,oldPVC=doesnot,featuregate=true; should keep field",
+			feature:  true,
+			spec:     withAllocatedResource("5G"),
+			oldSpec:  getPVCSpec(),
+			expected: withAllocatedResource("5G"),
+		},
+		{
+			name:     "for:newPVC=hasfield,oldPVC=hasfield,featuregate=false; should keep field",
+			feature:  false,
+			spec:     withAllocatedResource("10G"),
+			oldSpec:  withAllocatedResource("5G"),
+			expected: withAllocatedResource("10G"),
+		},
+		{
+			name:     "for:newPVC=hasfield,oldPVC=nil,featuregate=false; should drop field",
+			feature:  false,
+			spec:     withAllocatedResource("5G"),
+			oldSpec:  nil,
+			expected: getPVCSpec(),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.RecoverVolumeExpansionFailure, test.feature)()
+
+			DropDisabledFields(test.spec, test.oldSpec)
+
+			if !reflect.DeepEqual(*test.expected, *test.spec) {
+				t.Errorf("Unexpected change: %+v", diff.ObjectDiff(test.expected, test.spec))
+			}
+		})
+	}
+}
+
+func TestSetAllocatedResources(t *testing.T) {
+	tests := []struct {
+		name     string
+		feature  bool
+		spec     *core.PersistentVolumeClaimSpec
+		oldSpec  *core.PersistentVolumeClaimSpec
+		expected *core.PersistentVolumeClaimSpec
+	}{
+		{
+			name:     "feature:true; should default to requested size when crating new PVC",
+			feature:  true,
+			spec:     withResource(getPVCSpec(), "5G"),
+			oldSpec:  nil,
+			expected: withResource(withAllocatedResource("5G"), "5G"),
+		},
+		{
+			name:     "feature:true; should default to older allocated size if updated PVC is smaller",
+			feature:  true,
+			spec:     withResource(getPVCSpec(), "5G"),
+			oldSpec:  withResource(withAllocatedResource("10G"), "10G"),
+			expected: withResource(withAllocatedResource("10G"), "5G"),
+		},
+		{
+			name:     "feature:true; should default to higher allocated size if older allocated size is small",
+			feature:  true,
+			spec:     withResource(getPVCSpec(), "5G"),
+			oldSpec:  withResource(withAllocatedResource("4G"), "4G"),
+			expected: withResource(withAllocatedResource("5G"), "5G"),
+		},
+		{
+			name:     "feature:true; should default to older allocated size even if new allocated size is bigger",
+			feature:  true,
+			spec:     withResource(withAllocatedResource("30G"), "20G"),
+			oldSpec:  withResource(withAllocatedResource("20G"), "20G"),
+			expected: withResource(withAllocatedResource("20G"), "20G"),
+		},
+		{
+			name:     "feature:true; should default to older resource size if older resource size is bigger and old Allocatedresources is nil",
+			feature:  true,
+			spec:     withResource(withAllocatedResource("10G"), "20G"),
+			oldSpec:  withResource(getPVCSpec(), "40G"),
+			expected: withResource(withAllocatedResource("40G"), "20G"),
```

Comment:
Can you add test cases for allocatedResource>requestedResource in the oldSpec?

It seems impossible for allocatedResource < requestedResource in the oldSpec, but what if user messed it up with their own controller?  What will happen in that case?  Can you add test cases for those as well?

Reply:
Add a test case for `allocatedResources > requestedResource` for old spec. It should be impossible for `allocatedResources` to be smaller than `requestedResource` for oldSpec because these values are set by apiserver directly and is not part of any controller code and hence it should be impossible for user to mess it up.

Reply:
Ok